app/kuma-dp/pkg/dataplane/dnsproxy/component.go

```diff
@@ -84,12 +94,18 @@ func (s *Server) Handler(res dns.ResponseWriter, req *dns.Msg) {
 		response.Answer = append(response.Answer, dnsEntry.RR...)
 	} else {
 		proxyStart := time.Now()
-		c := new(dns.Client)
-		resp, _, err := c.Exchange(req, s.upstreamHostPort)
+		var resp *dns.Msg
+		var err error
+		if s.upstreamHandler != nil {
+			resp, err = s.upstreamHandler(req)
+		} else {
+			c := new(dns.Client)
+			resp, _, err = c.Exchange(req, s.upstreamHostPort)
+		}
```

Comment:
nitpick: we could just have an interface for `func(m *Msg, address string) (r *Msg, rtt time.Duration, err error)` and inject `dns.Client` or  a mock

Reply:
right reworked this way